### DISCUSSION TITLE
Error after upgrade to 22.11.1

### DISCUSSION THREAD

**Discussion by luddite516** [NetApp/harvest#1610] [Q&A] [answered]
I am in the process of upgrading Harvest from 21.11.0 to 22.11.1.

After installing the new packages and restarting services, I am seeing a new error that I do not understand:

`{"level":"error","Poller":"adcnapc07","exporter":"xxxxxxxxx","stack":"goroutine 14290 [running]:\ngithub.com/netapp/harvest/v2/pkg/logging.MarshalStack({0xbbf380?, 0xc000858230?})\n\tgithub.com/netapp/harvest/v2/pkg/logging/logger.go:157 +0x88\ngithub.com/rs/zerolog.(*Event).Err(0xc00e6000c0, {0xbbf380, 0xc000858230})\n\tgithub.com/rs/zerolog@v1.28.0/event.go:381 +0x63\ngithub.com/netapp/harvest/v2/cmd/exporters/prometheus.(*Prometheus).ServeMetrics(0xc0007ce8a0, {0xbc2d88, 0xc0000be0e0}, 0xc01ba60a00)\n\tgithub.com/netapp/harvest/v2/cmd/exporters/prometheus/httpd.go:132 +0x6fb\nnet/http.HandlerFunc.ServeHTTP(0xc0208aaaf0?, {0xbc2d88?, 0xc0000be0e0?}, 0x0?)\n\tnet/http/server.go:2109 +0x2f\nnet/http.(*ServeMux).ServeHTTP(0x0?, {0xbc2d88, 0xc0000be0e0}, 0xc01ba60a00)\n\tnet/http/server.go:2487 +0x149\nnet/http.serverHandler.ServeHTTP({0xc018e96720?}, {0xbc2d88, 0xc0000be0e0}, 0xc01ba60a00)\n\tnet/http/server.go:2947 +0x30c\nnet/http.(*conn).serve(0xc0001ca000, {0xbc3480, 0xc00086a9c0})\n\tnet/http/server.go:1991 +0x607\ncreated by net/http.(*Server).Serve\n\tnet/http/server.go:3102 +0x4db\n","error":"write tcp 127.0.0.1:12007->127.0.0.1:32962: write: broken pipe","caller":"prometheus/httpd.go:132","time":"2022-12-12T14:06:07-05:00","message":"write metrics"}
{"level":"error","Poller":"xxxxxxxxxx","exporter":"xxxxxxxxxxx","error":"write tcp 127.0.0.1:12007->127.0.0.1:32962: write: broken pipe","caller":"prometheus/httpd.go:137","time":"2022-12-12T14:06:07-05:00","message":"write ending newline"}`

I have not changed or updated anything else; just the Harvest RPM.

This particular poller collects a large number of metrics.  Could this be associated with a timeout exceeded?

There is one other poller we are running on this same host, and it is not displaying this error.  I collects fewer metrics than the one which is reporting the error.

**Comment by cgrinds** [comment]
hi @luddite516 looks like someone is scrapping metrics from the poller's Prometheus port and while Harvest is trying to write the response, the client gives up, dies, etc. and closes the connection. Harvest doesn't know the connection is closed yet and keeps writing, resulting in the `write: broken pipe`. The logging should be improved here, but the question is why is the client closing the connection before receiving all the data? Let's try to figure that out.

Few questions:

- Anything in the Harvest logs above that error?
- Is this poller being scraped by Prometheus? If so, anything in the Prometheus logs to correlate with this error?
- Does this always happen with this poller or intermittent?

Reference
- https://gosamples.dev/broken-pipe/

**Comment by luddite516** [comment, marked as answer]
> Anything in the Harvest logs above that error?

No, no other errors in Harvest logs.  Just info messages about collectors succeeding.

> Is this poller being scraped by Prometheus? If so, anything in the Prometheus logs to correlate with this error?

Yes Prometheus is scraping this target.  The prometheus log seems to be clean / not reporting any errors

> Does this always happen with this poller or intermittent?

Appears to be happening every 2 - 7 minutes, but not regularly.  

I suspect that something is timing out before it can read all the data.  Since I first posted my question, I have disabled some of the collectors by commenting them out of /opt/harvest/conf/zapiperf/default.yaml and restart the collectors.  Specifically, I disabled the two headroom collectors.  After restarting the collectors, the error has not appeared (last 10 minutes or so).
When I re-enable all collectors in default.yaml, the error comes back.

I will continue to test my assumptions and see if I can gain anymore insights.

Thanks for the quick response.

**Reply by cgrinds** [reply]
Thanks for the update @luddite516. I'll take a look at logging improvements in this area to get more information.

**Comment by luddite516** [comment]
Not knowing the exact cause of this issue, I tweaked a few parameters to see what would happen.  

I increased the prometheus scrape_timeout from 10s to 20s and restarted prometheus.  The error has not reappeared in the hour since I restarted.
I do not consider this to be a solution, becasue changing the scrape_timeout may have other negative effects.  But it shows that something is probably taking longer than it should.

Would you advise breaking up the prometheus scrape into two scrape jobs?

**Comment by luddite516** [comment]
Workaround for this issue:  increase the scrape_timeout specifically for the hosts's prometheus job from default 10s to 30s.  
Errors disappeared, and data collection was much more reliable / fewer missed scrapes.

In /etc/prometheus/prometheus.yml:

  - job_name: 'XXXXXXXX'
    static_configs:
      - targets:
        - 'localhost:12007'
    scrape_timeout: 30s

**Reply by cgrinds** [reply]
Thanks for the update @luddite516

I'm curious how many metrics the endpoint that's struggling has? Can you share the result of `curl http://localhost:12990/metrics | wc -l`?

Also what kind of `scrape_duration_seconds` times do you see for this troubled poller? Below is an example of that metric when I added a breakpoint in the exporter to artificially slow it down. The first peak at 10s is with the out-of-the-box `scrape_timeout`, which as you noted is 10s, the 2nd bump is when I changed it to 30s. What kind of response times are you seeing for this troubled poller?

<img width="797" alt="image" src="https://user-images.githubusercontent.com/242252/207697669-1d15773c-0820-4948-82ac-aed9963b40df.png">

**Comment by luddite516** [comment]
This endpoint contains almost 457,000 metrics

Over the past week, the scrape duration for this job has been as long as 100 sec, with an average of around 50 sec.

The endpoint is an OnTAP SnapVault cluster which holds around 8,000 volumes.  So that number of metrics is expected.
